lib/logstash/filters/jdbc/lookup.rb

```diff
@@ -60,44 +69,69 @@ def formatted_errors
+        result = if @column_validity_checked.true?
+                  LookupFailures.new
+                else
+                  @column_validity_checked.make_true
+                  LookupFailures.new.check_columns
+                end
+        local.fetch_with_lock(@id, event, result) # updates the instance of LookupFailures (see comments above), defined in JRuby extension
+        return true if result.successful?
+      rescue StandardError => e
         # a SQLException re-thrown as as a StandardError, the cause and backtrace has been logged already in Java
         # this error holds no meaningful info.
         tag_failure(event)
         return tag_and_set_default(event)
       end
 
+      if result.any_invalid_parameters?
+        logger.warn("Parameters for the statement cannot be prepared, interpolation may have failed or the value might be nil, array or hash", "parameters" => result.invalid_parameters, "event" => event.to_hash)
+        # if this is a checking columns run and there are invalid parameters then the prepared statement is not
+        # even executed so columns cannot validated this time - need to check columns again on next event.
+        @column_validity_checked.make_false if result.checking_columns?
+      else
+        if result.checking_columns? && result.any_invalid_columns?
+          # this means that a local db schema has been created with one or more datatypes
+          # that we can't coerce into acceptable Event datatypes. Other field/values are added to the Event.
+          invalid_columns = result.invalid_columns
+          all_columns = result.all_columns
+          @some_valid_columns = all_columns.size > invalid_columns.size
+          logger.error("The statement is returning data types that cannot be stored in an event", "invalid columns" => invalid_columns, "all columns" => all_columns)
+        end
+      end
+
       if result.invalid_id_for_lookup?
         logger.error("This local lookup has not been registered correctly, this is abnormal", "local_lookup id" => @id)
-        tag_failure(event)
-        return tag_and_set_default(event)
       end
 
-      if result.any_invalid_parameters?
-        logger.warn("Parameters for the statement cannot be prepared, interpolation may have failed or the value might be nil, array or hash", "parameters" => result.invalid_parameters, "event" => event.to_hash)
-        tag_failure(event)
-        return tag_and_set_default(event)
-      end
-      # this means that a local db schema has been created with one or more datatypes
-      # that we can't coerce into acceptable Event datatypes. Other field/values are added to the Event.
-      # Should we tag here or WARN only?
-      if result.any_invalid_columns?
-        logger.warn("The statement is returning datatypes that cannot be stored in an event", "invalid columns" => result.invalid_columns)
-        tag_failure(event)
-        return false
+      tag_failure(event)
+
+      if @some_valid_columns
```

Comment:
Are there any tests verifying this behavior, and with multiple events making sure that the same behaviour results when column checks are done and not?